### PR TITLE
CommandAvailabilityPublisher and websocket-based messaging

### PR DESCRIPTION
**Pull request by hatton**
Commands have a status that needs to be communicated to the html client. The CommandAvailabilityPublisher does that by wiring up to events on commands and sending updated status on commands over a websocket. This uses the "fleck" nuget library to implement websockets on the server side (they are built-in to Firefox on the client side).
This doesn't do anything the user can see yet. It's a bit of infrastructure that we will build on later. In particular, nothing on the javascript side is yet listening for these messages. Also, more work needs to be done about automatically wiring up Commands; see the comments in ProjectContext, where the suggested approaches didn't work, so only the DeletePage command is currently wired up, by hand.

### CODE REVIEW

**Comment by sillsdevgerrit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.lsdev.sil.org//job/GitHub-Bloom-Wrapper-debug/250/
Test PASSed.

**Comment by JohnThomson** [comment]
stats for status near the start of the commit message is a particularly confusing typo.

File: src/BloomExe/Command.cs [hunk 1]
```diff
@@ -12,43 +12,95 @@ public interface ICommand
...
+		}
 	}
 
-	public class CutCommand : Command { }
-	public class CopyCommand : Command { }
-	public class PasteCommand : Command{}
-	public class UndoCommand : Command { }
-	public class DuplicatePageCommand : Command { }
-	public class DeletePageCommand : Command { }
+	public class PasteCommand : Command
```
Comment (JohnThomson) [inline src/BloomExe/Command.cs]: It feels vaguely heavy-weight to have a whole class for each command, if they only differ by the value of one field. But there may be something about the framework that requires this.

File: src/BloomExe/ProjectContext.cs [hunk 2]
```diff
@@ -231,6 +232,18 @@ protected void BuildSubContainerForThisProject(string projectSettingsPath, ICont
 //					}
 					});
 				});
+
+				//TODO: why doesn't this work? It gives 0 commands
+				//_commandAvailabilityPublisher = new CommandAvailabilityPublisher(_scope.Resolve<IEnumerable<ICommand>>());
```
Comment (JohnThomson) [inline src/BloomExe/ProjectContext.cs]: Is it possible that we have to first tell autofac how to resolve a request for IEnumerable<ICommand>, presumably with something that can enumerate the one-and-only instance of all the concrete subclasses?

File: src/BloomExe/web/CommandAvailabilityPublisher.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Fleck;
+
+namespace Bloom.web
+{
+	/// <summary>
+	/// Commands have a stats that needs to be communicated to the html client.
```
Comment (JohnThomson) [inline src/BloomExe/web/CommandAvailabilityPublisher.cs]: stats->status

File: src/BloomExe/web/CommandAvailabilityPublisher.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		{
+			
+			foreach (var command in commands)
+			{
+				command.EnabledChanged += command_EnabledChanged;
+			}
+
+			FleckLog.Level = LogLevel.Warn;
+			_allSockets = new List<IWebSocketConnection>();
+			_server = new WebSocketServer("ws://127.0.0.1:8189");
```
Comment (JohnThomson) [inline src/BloomExe/web/CommandAvailabilityPublisher.cs]: Might be clearer, if there isn't some disadvantage, to use localhost instead of 127.0.0.1.

Comment (hatton) [inline src/BloomExe/web/CommandAvailabilityPublisher.cs]: localhost doesn't work with this library.

File: src/BloomExe/Command.cs (hunk 1 above)
Comment (hatton) [inline src/BloomExe/Command.cs]: I would call it verbose, but I tend to actually favor verbose over clever.

**Comment by sillsdevgerrit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.lsdev.sil.org//job/GitHub-Bloom-Wrapper-debug/253/
Test PASSed.

File: src/BloomExe/ProjectContext.cs (hunk 2 above)
Comment (JohnThomson) [inline src/BloomExe/ProjectContext.cs]: See http://stackoverflow.com/questions/8845604/autofac-resolveienumerablet-returns-empty-list.

Comment (JohnThomson) [inline src/BloomExe/ProjectContext.cs]: https://code.google.com/p/autofac/wiki/ImplicitCollectionSupport

Comment (hatton) [inline src/BloomExe/ProjectContext.cs]: Thanks. I've checked in a solution.

**Comment by sillsdevgerrit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.lsdev.sil.org//job/GitHub-Bloom-Wrapper-master-debug/309/
Test FAILed.

**Comment by hatton** [comment]
restest this please

**Comment by sillsdevgerrit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.lsdev.sil.org//job/GitHub-Bloom-Wrapper-master-debug/330/
Test PASSed.

**Comment by sillsdevgerrit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.lsdev.sil.org//job/GitHub-Bloom-Wrapper-master-debug/331/
Test PASSed.

**Pull request merged**